internal/provider/role_resource.go

```diff
@@ -0,0 +1,362 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	unleash "github.com/Unleash/unleash-server-api-go/client"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &roleResource{}
+	_ resource.ResourceWithConfigure   = &roleResource{}
+	_ resource.ResourceWithImportState = &roleResource{}
+)
+
+// NewRoleResource is a helper function to simplify the provider implementation.
+func NewRoleResource() resource.Resource {
+	return &roleResource{}
+}
+
+// roleResource is the resource implementation.
+type roleResource struct {
+	client *unleash.APIClient
+}
+
+type permissionRef struct {
+	Name        types.String `tfsdk:"name"`
+	Environment types.String `tfsdk:"environment"`
+}
+
+type roleResourceModel struct {
+	Id          types.Int64     `tfsdk:"id"`
+	Name        types.String    `tfsdk:"name"`
+	Type        types.String    `tfsdk:"type"`
+	Description types.String    `tfsdk:"description"`
+	Permissions []permissionRef `tfsdk:"permissions"`
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *roleResource) Configure(ctx context.Context, req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*unleash.APIClient)
```

Comment:
Okay now I'm even more confused by Golang! Surely this variable name should be `err`? And we should probably invert the check on the next line?

Reply:
Yeah... I think it's a design decision, but I'm also unsure about the syntax here...